### PR TITLE
Issue #24577: edit commonly-used-collection-types.md

### LINKED ISSUE #24577: Poor formatting and punctuation choices harm readability
# Issues
This article suffers from poor readability. I ran the page's content through a [Flesch-Kincaid readability test ](https://en.wikipedia.org/wiki/Flesch%E2%80%93Kincaid_readability_tests) and got a score of 37, which is poor. This indicates sentences and paragraphs that are too long, especially for ESL readers. More specific points:

- Many of the paragraphs are too large. This makes them feel like 'walls of text' that it is easy to get lost in.
- The large number of hyperlinked type-names concentrated in close proximity are hard to parse.
- Some phrases are needlessly convoluted, such as the first two sentences of the second paragraph.
- I feel that the use of semicolons in the final paragraph is inappropriate and should be replaced by bulletpoints.

# Suggestions
I would like to restructure and rephrase the article's content to make it easier to read without affecting any of the information it contains. More specifically:

- Bulletpoints should be used to clearly list different types.
- Paragraphs should be split up for easier reading. In particular, tag sentences (like 'Generic collections are the best solution to strong typing.') would more effectively frame the following content if they were their own paragraph.

# Example rewrite of the first two paragraphs

Collection types represent ways to collect data, such as hash tables, queues, stacks, bags, dictionaries, and lists.

All collections are based on the ICollection interface. IList, IDictionary and their generic counterparts are both derived from the ICollection interface. 

In collections which use IList or ICollection,  every element contains only a value. Examples include:

- Array
- ArrayList
- List<T>
- Queue
- ConcurrentQueue<T>
- Stack, ConcurrentStack<T>
- LinkedList<T> 

For collections which use IDictionary, every element contains both a key and a value. Examples include:

- Hashtable
- SortedList
- Dictionary<TKey,TValue>
- SortedList<TKey,TValue>
- ConcurrentDictionary<TKey,TValue>

The KeyedCollection<TKey,TItem> class is unique because it is a list of values with keys embedded within the values. Therefore, it behaves like a list and like a dictionary.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 64e986db-612f-2608-6949-26cd47881469
* Version Independent ID: 2e68de31-22ca-ac4c-cf85-4d452c9fbde0
* Content: [Commonly Used Collection Types](https://docs.microsoft.com/en-us/dotnet/standard/collections/commonly-used-collection-types#feedback)
* Content Source: [docs/standard/collections/commonly-used-collection-types.md](https://github.com/dotnet/docs/blob/main/docs/standard/collections/commonly-used-collection-types.md)
* Product: **dotnet-fundamentals**
* GitHub Login: @IEvangelist
* Microsoft Alias: **dapine**

### PR DESCRIPTION
**Pull request by jkendall327**
## Summary
Copy-edit for readability and flow by introducing more paragraph breaks, bulletpoints, headings and simplified language. I've also changed the date as this is a fairly major edit.

Please let me know if I've done anything wrong, I've not worked on the .NET docs before!

Fixes #24577

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/docs?pullRequest=24614) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/docs?pullRequest=24614) before we can accept your contribution.<br/><br/>:x: jkendall327 [sign now](https://cla.dotnetfoundation.org/dotnet/docs?pullRequest=24614)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/docs?pullRequest=24614) it.</sub>

File: docs/standard/collections/commonly-used-collection-types.md [hunk 1]
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
...
+The LINQ to Objects feature provides a common pattern for accessing the in-memory objects of any type that implements <xref:System.Collections.IEnumerable> or <xref:System.Collections.Generic.IEnumerable%601>. LINQ queries have several benefits over standard constructs like `foreach` loops:
+
+- They are concise and easier to understand
```
Comment (gewarren) [inline docs/standard/collections/commonly-used-collection-types.md:75]: Please add periods to the bullet points.

File: docs/standard/collections/commonly-used-collection-types.md [hunk 2]
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
...
+- They can filter, order and group data
```
Comment (gewarren) [inline docs/standard/collections/commonly-used-collection-types.md:76]:
```suggestion
- They can filter, order, and group data.
```

File: docs/standard/collections/commonly-used-collection-types.md [hunk 3]
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
...
+The LINQ to Objects feature provides a common pattern for accessing the in-memory objects of any type that implements <xref:System.Collections.IEnumerable> or <xref:System.Collections.Generic.IEnumerable%601>. LINQ queries have several benefits over standard constructs like `foreach` loops:
```
Comment (gewarren) [inline docs/standard/collections/commonly-used-collection-types.md:73]:
```suggestion
The LINQ to Objects feature provides a common pattern for accessing in-memory objects of any type that implements <xref:System.Collections.IEnumerable> or <xref:System.Collections.Generic.IEnumerable%601>. LINQ queries have several benefits over standard constructs like `foreach` loops:
```

File: docs/standard/collections/commonly-used-collection-types.md [hunk 4]
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
...
+- <xref:System.Collections.SortedList>  
+- <xref:System.Collections.Generic.SortedList%602> 
+- <xref:System.Collections.Generic.Dictionary%602>  
+- <xref:System.Collections.Concurrent.ConcurrentDictionary%602>  
+
+The <xref:System.Collections.ObjectModel.KeyedCollection%602> class is unique because it is a list of values with keys embedded within the values. As a result, it behaves both like a list and like a dictionary.  
+
+When you need efficient multi-threaded collection access, use the generic collections in the <xref:System.Collections.Concurrent> namespace.
+
+## Creating generic collections
```
Comment (gewarren) [inline docs/standard/collections/commonly-used-collection-types.md:45]:
```suggestion
## Strong typing
```

File: docs/standard/collections/commonly-used-collection-types.md [hunk 5]
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
...
+Generic collections are the best solution to strong typing. A `List<xref:System.Int32>`, for instance, will present a compile-time error if you add any other type to it.
```
Comment (gewarren) [inline docs/standard/collections/commonly-used-collection-types.md:46]:
```suggestion
Generic collections are the best solution to strong typing. For example, adding any type other than an <xref:System.Int32> to a `List<Int32>` collection causes a compile-time error.
```

File: docs/standard/collections/commonly-used-collection-types.md [hunk 6]
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
...
+ 
+If your language does not support generics, the <xref:System.Collections> namespace includes abstract base classes that you can extend to create strongly-typed collections:
```
Comment (gewarren) [inline docs/standard/collections/commonly-used-collection-types.md:48]:
```suggestion
However, if your language does not support generics, the <xref:System.Collections> namespace includes abstract base classes that you can extend to create collection classes that are strongly typed. These base classes include:
```

File: docs/standard/collections/commonly-used-collection-types.md [hunk 7]
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
...
+If your language does not support generics, the <xref:System.Collections> namespace includes abstract base classes that you can extend to create strongly-typed collections: 
+
+- <xref:System.Collections.CollectionBase>
+- <xref:System.Collections.ReadOnlyCollectionBase>
+- <xref:System.Collections.DictionaryBase>
+
+## How collections vary
   
- Generic collections are the best solution to strong typing. However, if your language does not support generics, the <xref:System.Collections> namespace includes base collections, such as <xref:System.Collections.CollectionBase>, <xref:System.Collections.ReadOnlyCollectionBase>, and <xref:System.Collections.DictionaryBase>, which are abstract base classes that can be extended to create collection classes that are strongly typed. When efficient multi-threaded collection access is required, use the generic collections in the <xref:System.Collections.Concurrent> namespace.  
+Collections vary depending on how they store elements, how they perform searches, how they compare elements, and how they are sorted.
```
Comment (gewarren) [inline docs/standard/collections/commonly-used-collection-types.md:56]:
```suggestion
Collections vary in how they store, sort, and compare elements, and how they perform searches. 
```

File: docs/standard/collections/commonly-used-collection-types.md [hunk 8]
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
...
+Collections vary depending on how they store elements, how they perform searches, how they compare elements, and how they are sorted. 
+
+All collections use zero-based indexes except <xref:System.Array>, which allows arrays that are not zero-based.  
+
+### Sorting
+You cannot sort the <xref:System.Collections.Hashtable> or the <xref:System.Collections.Generic.Dictionary%602> classes. The <xref:System.Collections.SortedList> class and the generic <xref:System.Collections.Generic.SortedList%602> provide sorted versions of these collections.
+
+The <xref:System.Collections.Queue> and <xref:System.Collections.Generic.Queue%601> classes provide first-in-first-out lists. The <xref:System.Collections.Stack> and <xref:System.Collections.Generic.Stack%601> classes provide last-in-first-out lists.
```
Comment (gewarren) [inline docs/standard/collections/commonly-used-collection-types.md:63]: This doesn't have anything to do with sorting.

File: docs/standard/collections/commonly-used-collection-types.md [hunk 9]
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
...
+
+## How collections vary
   
- Generic collections are the best solution to strong typing. However, if your language does not support generics, the <xref:System.Collections> namespace includes base collections, such as <xref:System.Collections.CollectionBase>, <xref:System.Collections.ReadOnlyCollectionBase>, and <xref:System.Collections.DictionaryBase>, which are abstract base classes that can be extended to create collection classes that are strongly typed. When efficient multi-threaded collection access is required, use the generic collections in the <xref:System.Collections.Concurrent> namespace.  
+Collections vary depending on how they store elements, how they perform searches, how they compare elements, and how they are sorted. 
+
+All collections use zero-based indexes except <xref:System.Array>, which allows arrays that are not zero-based.  
+
+### Sorting
+You cannot sort the <xref:System.Collections.Hashtable> or the <xref:System.Collections.Generic.Dictionary%602> classes. The <xref:System.Collections.SortedList> class and the generic <xref:System.Collections.Generic.SortedList%602> provide sorted versions of these collections.
```
Comment (gewarren) [inline docs/standard/collections/commonly-used-collection-types.md:61]: I don't think that's true. You can call OrderBy to sort dictionaries and hash tables. I would just keep the original version of this sentence.

File: docs/standard/collections/commonly-used-collection-types.md [hunk 10]
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
 ---
 # Commonly Used Collection Types
```
Comment (gewarren) [inline docs/standard/collections/commonly-used-collection-types.md:16]:
```suggestion
# Commonly used collection types
```

File: docs/standard/collections/commonly-used-collection-types.md [hunk 11]
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
...
+Collection types represent ways to collect data, such as hash tables, queues, stacks, bags, dictionaries, and lists.
```
Comment (gewarren) [inline docs/standard/collections/commonly-used-collection-types.md:18]:
```suggestion
Collection types represent different ways to collect data, such as hash tables, queues, stacks, bags, dictionaries, and lists.
```

File: docs/standard/collections/commonly-used-collection-types.md [hunk 12]
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
...
+
+All collections are based on the <xref:System.Collections.ICollection> interface. <xref:System.Collections.IList>, <xref:System.Collections.IDictionary> and their generic counterparts both derive from <xref:System.Collections.ICollection>.
```
Comment (gewarren) [inline docs/standard/collections/commonly-used-collection-types.md:20]:
```suggestion
All collections are based on the <xref:System.Collections.ICollection> or <xref:System.Collections.Generic.ICollection%601> interface, either directly or indirectly. <xref:System.Collections.IList> and <xref:System.Collections.IDictionary> and their generic counterparts all derive from these two interfaces. 
```

File: docs/standard/collections/commonly-used-collection-types.md [hunk 13]
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
...
+Collection types represent ways to collect data, such as hash tables, queues, stacks, bags, dictionaries, and lists.
+
+All collections are based on the <xref:System.Collections.ICollection> interface. <xref:System.Collections.IList>, <xref:System.Collections.IDictionary> and their generic counterparts both derive from <xref:System.Collections.ICollection>. 
+
+In collections based on <xref:System.Collections.IList>, or directly on <xref:System.Collections.ICollection>, every element contains only a value. Examples include:
```
Comment (gewarren) [inline docs/standard/collections/commonly-used-collection-types.md:22]:
```suggestion
In collections based on <xref:System.Collections.IList> or directly on <xref:System.Collections.ICollection>, every element contains only a value. These types include:
```

File: docs/standard/collections/commonly-used-collection-types.md [hunk 14]
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
...
+- <xref:System.Array>
+- <xref:System.Collections.ArrayList>
+- <xref:System.Collections.Generic.List%601> 
+- <xref:System.Collections.Queue>
+- <xref:System.Collections.Concurrent.ConcurrentQueue%601>
+- <xref:System.Collections.Stack>
+- <xref:System.Collections.Concurrent.ConcurrentStack%601>
+- <xref:System.Collections.Generic.LinkedList%601>
+
+In collections based on the <xref:System.Collections.IDictionary> interface, every element contains both a key and a value. Examples include:
```
Comment (gewarren) [inline docs/standard/collections/commonly-used-collection-types.md:33]:
```suggestion
In collections based on the <xref:System.Collections.IDictionary> interface, every element contains both a key and a value. These types include:
```

**Review by gewarren — CHANGES_REQUESTED** [review]
@jkendall327 thanks for your contribution to improving docs. I've left some requested changes.

**Comment by jkendall327** [comment]
@gewarren Thanks for the comments, I've just incorporated them all. Let me know if you think any further changes would help.

**Comment by gewarren** [comment]
I made a few more changes directly, hope you don't mind. Looks like there are still some trailing spaces to clean up.

For the periods on the bullet points, that's only if the bullet points are more than a few words. If it's just one word, don't add a period.

**Comment by jkendall327** [comment]
No worries, I appreciate the improvements. I've just removed the trailing spaces caught by the linter. Thanks!

**Review by gewarren — APPROVED** [review]

**Comment by gewarren** [comment]
This looks good to me. @IEvangelist do you want to sign off on this? [Preview link](https://review.docs.microsoft.com/en-us/dotnet/standard/collections/commonly-used-collection-types?branch=pr-en-us-24614).

**Review by IEvangelist — APPROVED** [review]
Looks good to me, we'll :shipit: - thank you 🙏
